broker-daemon/block-order-worker/index.js

```diff
@@ -316,6 +281,43 @@ class BlockOrderWorker extends EventEmitter {
     }
   }
 
+  /**
+   * Move a block order to a completed state if all orders have been completed
+   * @todo What should we do if this method fails, but the order itself is completed?
+   * @param {String} blockOrderId
+   */
+  async completeBlockOrder (blockOrderId) {
+    this.logger.info('Attempting to put block order in a completed state', { id: blockOrderId })
+
+    const blockOrder = await this.getBlockOrder(blockOrderId)
+
+    // check the fillamount on each collection of state machines from the block order
+    // and make sure that either is equal to how much we are trying to fill.
+    let totalFilled = Big(0)
+    totalFilled = blockOrder.fills.reduce((acc, fsm) => acc.plus(fsm.fill.fillAmount), totalFilled)
+    totalFilled = blockOrder.openOrders.reduce((acc, osm) => {
+      // If the order has not been filled yet, then the `fillAmount` will be undefined
+      // so we instead default to 0
+      return acc.plus(osm.order.fillAmount || 0)
+    }, totalFilled)
+
+    this.logger.debug('Current total filled amount: ', { totalFilled, blockOrderAmount: blockOrder.baseAmount })
+
+    const stillBeingFilled = totalFilled.lt(blockOrder.baseAmount)
+    const activeOrders = blockOrder.openOrders.some(osm => osm.order.active)
+    const activeFills = blockOrder.fills.some(fsm => fsm.fill.active)
+
+    if (!activeOrders && !activeFills && !stillBeingFilled) {
+      // An order is only completed if all orders underneath the blockorder are out of
+      // an `ACTIVE` state and it is entirely filled
+      blockOrder.complete()
+      await promisify(this.store.put)(blockOrder.key, blockOrder.value)
+      this.logger.info('Moved block order to completed state', { blockOrderId })
+    } else {
+      this.logger.debug('Block order is not ready to be completed', { blockOrderId })
```

Comment:
should this be an error?

Reply:
Not necessarily. There are cases where this would be an error, however we decided to remove them. As an example: This would be hit if the block order is still being filled, which is a valid case and wouldn't be an error